File: python/src/swiftpay/modules/x402_facilitator.py
"""x402 facilitator endpoints. No authentication required — protocol is signature-bound."""

from __future__ import annotations

from typing import TYPE_CHECKING

from swiftpay.types import (
    FacilitatorRequest,
    PaymentStatusResponse,
    SettlementResponse,
    SupportedResponse,
    VerifyResponse,
)

if TYPE_CHECKING:
    from swiftpay.async_client import AsyncHTTPClient
    from swiftpay.client import HTTPClient

# OpenAPI returns SettlementResponse for both 200 (success) and 422 (settlement
# failed). settle() inspects the body for both — other 4xx/5xx still raise.
_SETTLE_ACCEPT = (422,)


def _serialise_facilitator_request(request: FacilitatorRequest) -> dict[str, object]:
    return request.model_dump(by_alias=True, exclude_none=True)


# ── Sync ─────────────────────────────────────────────────────────────────────


class X402FacilitatorModule:
    def __init__(self, http: HTTPClient) -> None:
        self._http = http

    def supported(self) -> SupportedResponse:
        """``GET /v1/x402/supported``."""
        data = self._http.request("GET", "/v1/x402/supported")
        return SupportedResponse.model_validate(data)

    def verify(self, request: FacilitatorRequest) -> VerifyResponse:
        """``POST /v1/x402/verify`` — validate without on-chain side effects."""
        data = self._http.request(
            "POST", "/v1/x402/verify", json_body=_serialise_facilitator_request(request)
        )
        return VerifyResponse.model_validate(data)

    def settle(self, request: FacilitatorRequest) -> SettlementResponse:
        """``POST /v1/x402/settle`` — execute the on-chain transfer.

        Returns ``SettlementResponse`` for both 200 and 422; inspect ``success``
        / ``error`` to decide. Other 4xx/5xx raise.
        """
        raw = self._http.request_raw(
            "POST",
            "/v1/x402/settle",
            json_body=_serialise_facilitator_request(request),
            accept_statuses=_SETTLE_ACCEPT,
        )
        return SettlementResponse.model_validate(raw.body)

    def payment_status(self, *, nonce: str) -> PaymentStatusResponse:
        """``GET /v1/x402/payments/status?nonce=...``."""
        data = self._http.request("GET", "/v1/x402/payments/status", query={"nonce": nonce})
        return PaymentStatusResponse.model_validate(data)


# ── Async ────────────────────────────────────────────────────────────────────


class AsyncX402FacilitatorModule:
    def __init__(self, http: AsyncHTTPClient) -> None:
        self._http = http

    async def supported(self) -> SupportedResponse:
        data = await self._http.request("GET", "/v1/x402/supported")
        return SupportedResponse.model_validate(data)

    async def verify(self, request: FacilitatorRequest) -> VerifyResponse:
        data = await self._http.request(
            "POST", "/v1/x402/verify", json_body=_serialise_facilitator_request(request)
        )
        return VerifyResponse.model_validate(data)

    async def settle(self, request: FacilitatorRequest) -> SettlementResponse:
        raw = await self._http.request_raw(
            "POST",
            "/v1/x402/settle",
            json_body=_serialise_facilitator_request(request),
            accept_statuses=_SETTLE_ACCEPT,
        )
        return SettlementResponse.model_validate(raw.body)

    async def payment_status(self, *, nonce: str) -> PaymentStatusResponse:
        data = await self._http.request("GET", "/v1/x402/payments/status", query={"nonce": nonce})
        return PaymentStatusResponse.model_validate(data)

File: nodejs/src/modules/x402-facilitator.ts
import type { HTTPClient } from '../client.js';
import type {
  FacilitatorRequest,
  PaymentStatusResponse,
  SettlementResponse,
  SupportedResponse,
  VerifyResponse,
} from '../types.js';

/**
 * x402 facilitator endpoints. No authentication required — the protocol carries
 * its own EIP-712 signatures end-to-end.
 */
export class X402FacilitatorModule {
  constructor(private readonly http: HTTPClient) {}

  /** GET /v1/x402/supported — list supported network/scheme combinations. */
  async supported(): Promise<SupportedResponse> {
    return this.http.request<SupportedResponse>({
      method: 'GET',
      path: '/v1/x402/supported',
    });
  }

  /** POST /v1/x402/verify — validate a payment payload without on-chain side effects. */
  async verify(body: FacilitatorRequest): Promise<VerifyResponse> {
    return this.http.request<VerifyResponse>({
      method: 'POST',
      path: '/v1/x402/verify',
      body,
    });
  }

  /**
   * POST /v1/x402/settle — execute the on-chain transfer.
   *
   * The OpenAPI spec returns `SettlementResponse` for both 200 (success) and
   * 422 (settlement failed). This method returns the body for both — inspect
   * `success` / `error` to decide. All other 4xx/5xx still throw.
   */
  async settle(body: FacilitatorRequest): Promise<SettlementResponse> {
    const raw = await this.http.requestRaw<SettlementResponse>({
      method: 'POST',
      path: '/v1/x402/settle',
      body,
      acceptStatuses: [422],
    });
    return raw.body;
  }

  /** GET /v1/x402/payments/status — look up a payment by its EIP-3009 nonce. */
  async paymentStatus(params: { nonce: string }): Promise<PaymentStatusResponse> {
    return this.http.request<PaymentStatusResponse>({
      method: 'GET',
      path: '/v1/x402/payments/status',
      query: { nonce: params.nonce },
    });
  }
}
